File: AI_Chat_Sever/AI_Chat_Sever/DialogueService.cpp
#include "DialogueService.h"
#include "DialoguePolicy.h"
#include "DialogueCache.h"
#include "LLMClient.h"

DialogueResult DialogueService::GenerateDialogue(
    const CharacterState& state,
    const DialogueSituation& situation)
{
    DialoguePolicy policy;
    DialogueCache cache;

    std::string key = policy.MakeCacheKey(state, situation);

    if (cache.Exists(key))
        return cache.Get(key);

    if (!policy.AllowLLMCall(state, situation))
        return policy.GetFallbackDialogue(state);

    LLMClient llm;
    DialogueResult result = llm.GenerateDialogue(state, situation);

    cache.Store(key, result);
    return result;
}

File: AI_chat_Sever_PY/ai_service/service/dialogue_service.py
from policy.dialogue_policy import allow_llm_call, make_cache_key, fallback_dialogue
from cache.dialogue_cache import get_cache, set_cache
from llm.llm_client import call_llm

async def generate_dialogue(req):
    key = make_cache_key(req)

    cached = get_cache(key)
    if cached:
        return cached

    if not allow_llm_call(req):
        return fallback_dialogue(req)

    result = await call_llm(req)
    set_cache(key, result)
    return result